### PR TITLE
MMScan devkit

### PR DESCRIPTION
**Pull request by rbler1234**
Thanks for your contribution and we appreciate it a lot. The following instructions would make your pull request more healthy and more easily get feedback. If you do not understand some items, don't worry, just make the pull request and seek help from maintainers.

## Modification

Update the branch regarding MMScan.

## Modification
- MMScan data processing code.
- Introduction and tutorial for MMScan.
- MMScan's data loading and evaluation-related API code.
(The model-related code will be uploaded subsequently.）

## BC-breaking (Optional)

Does the modification introduce changes that break the back-compatibility of the downstream repos?
If so, please describe how it breaks the compatibility and how the downstream projects should modify their code to keep compatibility with this PR.

## Use cases (Optional)

If this PR introduces a new feature, it is better to list some use cases here, and update the documentation.

## Checklist

1. Pre-commit or other linting tools are used to fix the potential lint issues.
2. The modification is covered by complete unit tests. If not, please add more unit test to ensure the correctness.
3. If the modification has potential influence on downstream projects, this PR should be tested with downstream projects.
4. The documentation has been modified accordingly, like docstring or example tutorials.

### CODE REVIEW

File: data_preparation/process_all_scan.py [hunk 1]
```diff
@@ -0,0 +1,229 @@
...
+
+        box_pc_mask = is_inside_box(pc, center, size, orientation)
+
+        instance_ids[box_pc_mask] = obj_id
+        label[box_pc_mask] = obj_type_id
+    return pc, color, label, instance_ids
+
+
+@mmengine_track_func
+def process_one_scan(
```
Comment (mxh1999) [inline data_preparation/process_all_scan.py:76]: type of args

File: data_preparation/process_all_scan.py [hunk 2]
```diff
@@ -0,0 +1,229 @@
...
+    The pth file is a tuple of:
+        (1) aliged point clouds coordinates
+            shape (n,3)
+        (2) point clouds color ([0,1])
+            shape (n,3)
+        (3) point clouds label (int)
+            shape (n,1)
+        (4) point clouds object id (int)
+            shape (n,1)
+    Args:
```
[the remark this answers is not in the document]
Comment (mxh1999) [reply data_preparation/process_all_scan.py:98]: doc strings follow embodiedscan

File: data_preparation/utils/scannet_process.py [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+                     scan_id + '_vh_clean_2.ply'))
+    data = PlyData.read(scan_ply_path)
+    x = np.asarray(data.elements[0].data['x']).astype(np.float32)
+    y = np.asarray(data.elements[0].data['y']).astype(np.float32)
+    z = np.asarray(data.elements[0].data['z']).astype(np.float32)
+    pc = np.stack([x, y, z], axis=1)
+    r = np.asarray(data_color.elements[0].data['red'])
+    g = np.asarray(data_color.elements[0].data['green'])
+    b = np.asarray(data_color.elements[0].data['blue'])
+    pc_color = (np.stack([r, g, b], axis=1) / 256.0).astype(np.float32)
```
Comment (mxh1999) [inline data_preparation/utils/scannet_process.py:21]: /256 or /255?

File: mmscan/evaluator/gpt_evaluation.py [hunk 4]
```diff
@@ -0,0 +1,252 @@
...
+                     user_content_groups,
+                     max_tokens=1000):
+        """Calling the GPT api, return the results in the format of json.
+
+        Args:
+            system_prompt (str) :
+                The system prompt inputted into GPT.
+            user_content_grounps (list[str]) :
+                The user content inputted into GPT.
+            max_tokens (int) : max tokens, default 1000.
```
Comment (mxh1999) [inline mmscan/evaluator/gpt_evaluation.py:56]: Defaults to 1000

File: data_preparation/process_all_scan.py (hunk 1 above)
Comment (mxh1999) [reply data_preparation/process_all_scan.py:76]: Also check other places

File: mmscan/evaluator/metrics/box_metric.py [hunk 5]
```diff
@@ -0,0 +1,239 @@
...
+
+def average_precision(recalls, precisions, mode='area'):
+    """Calculate average precision (for single or multiple scales).
+
+    Args:
+        recalls (np.ndarray): Recalls with shape of (num_scales, num_dets)
+            or (num_dets, ).
+        precisions (np.ndarray): Precisions with shape of
+            (num_scales, num_dets) or (num_dets, ).
+        mode (str): 'area' or '11points', 'area' means calculating the area
```
Comment (mxh1999) [inline mmscan/evaluator/metrics/box_metric.py:13]: Defaults to

File: data_preparation/process_all_scan.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (mxh1999) [reply data_preparation/process_all_scan.py:98]: also check other places

File: mmscan/evaluator/vg_evaluation.py [hunk 6]
```diff
@@ -0,0 +1,329 @@
...
+
+        category_records(dict): Metric results for each category
+            (average of all samples with the same category)
+    Args:
+        verbose(bool): Whether to print the evaluation results.
+            Defaults to True.
+    """
+
+    def __init__(self, verbose=True) -> None:
+        print('new methods!')
```
Comment (mxh1999) [inline mmscan/evaluator/vg_evaluation.py:30]: remove unnecessary print

File: mmscan/evaluator/vg_evaluation.py [hunk 7]
```diff
@@ -0,0 +1,329 @@
...
+                                                 get_average_precision,
+                                                 get_multi_topk_scores)
+from mmscan.utils.box_utils import index_box, to_9dof_box
+
+
+class VG_Evaluator:
+    """Evaluator for MMScan Visual Grounding benchmark.
+
+    Attributes:
+        eval_metric: All the evaluation metric, includes
```
Comment (mxh1999) [inline mmscan/evaluator/vg_evaluation.py:16]: args are hard coded?
And the docstring does not match the default value in the code

File: mmscan/mmscan.py [hunk 8]
```diff
@@ -0,0 +1,509 @@
...
+import sys
+import time
+from copy import deepcopy
+from typing import List
+
+import numpy as np
+import torch
+from torch.utils.data import Dataset
+
+from mmscan.utils.box_utils import __9dof_to_6dof__
```
Comment (mxh1999) [inline mmscan/mmscan.py:12]: you cannot import __xxxxx from other module

File: models/README.md [hunk 9]
```diff
@@ -0,0 +1,119 @@
+## Visual Grounding Models
```
Comment (mxh1999) [inline models/README.md:1]: Titles and descriptions.

**Review by mxh1999 — CHANGES_REQUESTED** [review]
I think the structure is a little strange.
Maybe it would be better not to be a branch of embodiedscan

File: mmscan/utils/lang_utils.py [hunk 10]
```diff
@@ -0,0 +1,236 @@
+# flake8: noqa
+import re
+import string
+
+
```
[the remark this answers is not in the document]
Comment (mxh1999) [reply mmscan/utils/lang_utils.py:5]: # refer to LEO: embodied-generalist
# https://github.com/embodied-generalist/embodied-generalist/blob/477dc44b8b18dbfbe6823c307436d896ec8b062e/data/data_utils.py#L322-L379

[the remark this answers is not in the document]
Comment (mxh1999) [reply mmscan/utils/lang_utils.py:5]: `# refer to LEO: embodied-generalist`
`# https://github.com/embodied-generalist/embodied-generalist/blob/477dc44b8b18dbfbe6823c307436d896ec8b062e/data/data_utils.py#L322-L379`

**Review by mxh1999 — COMMENTED** [review]
I think the code (excluding the third-party model) is ok

**Review by mxh1999 — APPROVED** [review]

**Comment by mxh1999** [comment]
@rbler1234 Don't modify this branch until it has been merged, you can create another branch for develop.

File: assets/2024_NeurIPS_MMScan_Camera_Ready.pdf
Comment (Tai-Wang) [inline assets/2024_NeurIPS_MMScan_Camera_Ready.pdf:1]: It is better to compress the pdf to make it smaller

File: data_preparation/README.md [hunk 11]
```diff
@@ -0,0 +1,51 @@
...
+1. Download ScanNet v2 data [HERE](https://github.com/ScanNet/ScanNet). Link or move the folder to this level of directory.
+
+2. Download 3RScan data [HERE](https://github.com/WaldJohannaU/3RScan). Link or move the folder to this level of directory.
+
+3. Download Matterport3D data [HERE](https://github.com/niessner/Matterport). Link or move the folder to this level of directory.
+
+4. Organize the file structure. You are recommanded to create a soft link to the raw data folder under `mmsan_data/embodiedscan-split/data`.
+
+   ```
+   mmsan_data/embodiedscan-split/data/
```
Comment (Tai-Wang) [inline data_preparation/README.md:15]: typo, mmscan, please also fix other places

Comment (Tai-Wang) [reply data_preparation/README.md:15]: Why do we need the embodiedscan-split folder

File: data_preparation/utils/data_utils.py [hunk 12]
```diff
@@ -0,0 +1,130 @@
...
+    object_int_to_type = {v: k for k, v in object_type_to_int.items()}
+    datalist = data['data_list']
+    output_data = {}
+    pbar = (tqdm(range(len(datalist))) if show_progress else range(
+        len(datalist)))
+    for scene_idx in pbar:
+        # print(datalist[scene_idx]['sample_idx'])
+        # if "matterport3d" not in datalist[scene_idx]['sample_idx']:
+        #     continue
+        # print(datalist[scene_idx].keys())
```
Comment (Tai-Wang) [inline data_preparation/utils/data_utils.py:45]: remove unused codes

File: data_preparation/utils/mp3d_process.py [hunk 13]
```diff
@@ -0,0 +1,36 @@
+import numpy as np
+from plyfile import PlyData
+
+
+def process_mp3d(new_scan_id: str, data_root: str,
+                 axis_align_matrix_dict: dict, mapping: dict):
+    """Process matterport3d data.
+
+    Args:
+        scan_id (str): ID of the matterport3d scan.
```
Comment (Tai-Wang) [inline data_preparation/utils/mp3d_process.py:10]: scan_id in docstring is inconsistent with args new_scan_id

File: data_preparation/utils/pcd_utils.py [hunk 14]
```diff
@@ -0,0 +1,130 @@
...
+                  rotation_mat: np.ndarray) -> np.ndarray:
+    """Check if points are inside a 3D bounding box.
+
+    Args:
+        points(np.ndarray): 3D points, numpy array of shape (n, 3).
+        center(np.ndarray): center of the box, numpy array of shape (3, ).
+        size(np.ndarray): size of the box, numpy array of shape (3, ).
+        rotation_mat(np.ndarray): rotation matrix of the box,
+            numpy array of shape (3, 3).
+    Returns:
```
Comment (Tai-Wang) [inline data_preparation/utils/pcd_utils.py:41]: need a blank line before Returns

File: data_preparation/utils/pcd_utils.py [hunk 15]
```diff
@@ -0,0 +1,130 @@
...
+    matrices = [x.squeeze(axis=-3) for x in matrices]
+    return np.matmul(np.matmul(matrices[0], matrices[1]), matrices[2])
+
+
+def euler_to_matrix_np(euler):
+    """Convert rotations given as Euler angles in radians to rotation matrices.
+
+    Args:
+        euler (np.ndarray) : (..., 3)
+    Returns:
```
Comment (Tai-Wang) [inline data_preparation/utils/pcd_utils.py:124]: blank line

File: data_preparation/utils/proc_utils.py [hunk 16]
```diff
@@ -0,0 +1,11 @@
+from functools import wraps
+
+
+def mmengine_track_func(func):
```
Comment (Tai-Wang) [inline data_preparation/utils/proc_utils.py:4]: Can we directly use the function in mmengine?

File: data_preparation/utils/trscan_process.py [hunk 17]
```diff
@@ -0,0 +1,44 @@
+import os
+
+import numpy as np
+import torch
+from pytorch3d.io import load_obj
+
+
+def process_trscan(scan_id: str, data_root: str, axis_align: dict):
```
Comment (Tai-Wang) [inline data_preparation/utils/trscan_process.py:8]: trscan? -> 3rscan?

File: mmscan/__init__.py [hunk 18]
```diff
@@ -0,0 +1,16 @@
+# flake8: noqa
+from mmscan.mmscan import MMScan
+
+print('MMScan module loaded')
+try:
+    from mmscan.evaluator.vg_evaluation import VG_Evaluator
```
Comment (Tai-Wang) [inline mmscan/__init__.py:6]: naming convention, we would like prefer the class name like VG_Evaluator -> VisualGroundingEvaluator
Please also fix other names

File: data_preparation/process_all_scan.py [hunk 19]
```diff
@@ -0,0 +1,243 @@
...
+from utils.data_utils import read_annotation_pickle
+from utils.mp3d_process import process_mp3d
+from utils.pcd_utils import is_inside_box
+from utils.scannet_process import process_scannet
+from utils.trscan_process import process_3rscan
+
+es_anno = {}
+
+
+def mmengine_track_func(func):
```
Comment (Tai-Wang) [inline data_preparation/process_all_scan.py:21]: Is this function necessary?

Comment (rbler1234) [reply data_preparation/process_all_scan.py:21]: It is necessary if using the 'mmengine'  to run multi-process multi-parameter functions.

File: mmscan/evaluator/gpt_evaluation.py [hunk 20]
```diff
@@ -0,0 +1,260 @@
...
+    """GPT metric, we set this for QA and Caption tasks.
+
+    Args:
+        eval_size (int) : The number of samples to evaluate, -1 means
+            all samples.
+            Defaults to -1.
+        api_key (str) : The openai key.
+        model (str) : The GPT model to use, default we use "gpt-4o-mini".
+            Defaults to "gpt-4o-mini".
+        verbose (bool) : Whether to print the evaluation results or not.
```
Comment (Tai-Wang) [inline mmscan/evaluator/gpt_evaluation.py:22]: inconsistent with args

File: mmscan/evaluator/gpt_evaluation.py [hunk 21]
```diff
@@ -0,0 +1,260 @@
...
+                tmp-stored json files.
+        """
+
+        system_prompt, ex_instance = qa_prompt_define()
+
+        # Define the number of retries
+        MAXTRY = 3
+        gpt_eval_results = {}
+
+        for QA_ID in tqdm(QA_sample_dict):
```
Comment (Tai-Wang) [inline mmscan/evaluator/gpt_evaluation.py:95]: qa_id, qa_sample_dict

File: mmscan/evaluator/metrics/box_metric.py [hunk 22]
```diff
@@ -0,0 +1,277 @@
...
+    precision = tp_cum / np.maximum(tp_cum + fp_cum, np.finfo(np.float64).eps)
+
+    return average_precision(recall, precision), np.max(recall)
+
+
+def get_multi_topk_scores(iou_array: Union[np.ndarray, torch.tensor],
+                          iou_threshold: float,
+                          mode: str = 'sigma') -> Dict[str, float]:
+    """
+    Compute the multi-topk metric, we provide two modes.
```
Comment (Tai-Wang) [inline mmscan/evaluator/metrics/box_metric.py:229]: move it 1 line above

File: mmscan/evaluator/metrics/box_metric.py [hunk 23]
```diff
@@ -0,0 +1,277 @@
...
+    "simple": 1/N *Hit(min(N*k,len(pred)))
+    "sigma": 1/N * Sigma [Hit(min(n*k,len(pred)))>=n] n = 1~N
+        Hit(M) return the number of gtound truths hitted by the first
+        M predictions.
+        N = the number of gtound truths
```
Comment (Tai-Wang) [inline mmscan/evaluator/metrics/box_metric.py:234]: can move the explanation about different modes to the docstring about the "argument mode"

File: mmscan/evaluator/metrics/lang_metric.py [hunk 24]
```diff
@@ -0,0 +1,299 @@
...
+            caps = kvs[k]
+            for c in caps:
+                res[k].append({'caption': c})
+        else:
+            res[k].append({'caption': ''})
+    return res
+
+
+def coco_evaluate(batch_input: List[dict]) -> Tuple[dict, dict]:
+    """Calculate the extract matching score for each item.
```
Comment (Tai-Wang) [inline mmscan/evaluator/metrics/lang_metric.py:29]: need a blank line below

File: mmscan/evaluator/metrics/lang_metric.py [hunk 25]
```diff
@@ -0,0 +1,299 @@
...
+                final_list[m] = s_
+        else:
+            final_scores[method] = score
+            final_list[method] = scores
+
+    return final_scores, final_list
+
+
+def em_evaluation(batch_input: List[dict]) -> Tuple[list, list]:
+    """Calculate the extract matching score for each item.
```
Comment (Tai-Wang) [inline mmscan/evaluator/metrics/lang_metric.py:77]: Please also fix the blank line here

File: mmscan/evaluator/metrics/lang_metric.py [hunk 26]
```diff
@@ -0,0 +1,299 @@
...
+                correct = 1
+                continue
+            elif ''.join(gt.split()) in ''.join(pred.split()):
+                correct = 1
+                continue
+        refine_em_result.append(correct)
+    return em_result, refine_em_result
+
+
+class simcse_evaluator:
```
Comment (Tai-Wang) [inline mmscan/evaluator/metrics/lang_metric.py:122]: SimCSEEvalator

File: mmscan/evaluator/metrics/lang_metric.py [hunk 27]
```diff
@@ -0,0 +1,299 @@
...
+                             gt_count: List[int]) -> List[float]:
+        """Using Sentence Embeddings to calculate similarity between pred/gt in
+        a batch.
+
+        Args:
+            gt_count(list[int]):
+                stores number of possible answers to a question
+            all_pred(list[str]): all prediction
+            all_gt(list[str]): all ground truth
+            tips: len(all_gt)>=len(all_pred)
```
Comment (Tai-Wang) [inline mmscan/evaluator/metrics/lang_metric.py:146]: Move to the beginning

File: mmscan/evaluator/metrics/lang_metric.py [hunk 28]
```diff
@@ -0,0 +1,299 @@
...
+        self.simcse_tokenizer = AutoTokenizer.from_pretrained(model_path)
+        self.simcse_model = AutoModel.from_pretrained(model_path).to('cuda')
+
+    def __batch_evaluation__(self, all_pred: List[str], all_gt: List[str],
+                             gt_count: List[int]) -> List[float]:
+        """Using Sentence Embeddings to calculate similarity between pred/gt in
+        a batch.
+
+        Args:
+            gt_count(list[int]):
```
Comment (Tai-Wang) [inline mmscan/evaluator/metrics/lang_metric.py:142]: The order of arguments is inconsistent between the function call and docstring
